### ISSUE TITLE
Sync libraries and binaries from AUR before the package depending on them

### ISSUE THREAD

**Issue by gS644** [archlinuxfr/yaourt#186]
For example, today I got:

```
build@BlueArch ~> yaourt -Syua
:: Synchronizing package databases...
 core is up to date
 extra is up to date
 community is up to date
 multilib is up to date
 Foreign packages: - 65 / 65

==> Software upgrade (new version) :
aur/python3-aur    2015.12.6.4-1            -> 2015.12.7-1
aur/python3-xcgf   2015.12-1                -> 2015.12.7.1-1
```

but `python3-aur` is synced before `python3-xcgf`, which is a dependency of `python3-aur`

**Comment by f2404** [comment]
Hi @nullptrT 
It's a bit tricky as the packages are just sorted by their names currently...

Can you tell how ```pacman``` handles such cases? does it install the dependencies first?

**Comment by nullptrT** [comment]
Probably by some sort of a pactree, don't really know. This may be too difficult to implement for shell scripts...

You may classify this as "needed, but optional" (sounds weird, I know)

**Comment by hauptmech** [comment]
With the recent ABI change I needed to rebuild a large number of AUR packages in dependency order. After searching a while for an existing solution [I made my own](https://gist.github.com/hauptmech/19eae1b1a54e2eb2791b). I don't have time to make a patch for yaourt but you can see from the gist that it should be pretty easy to add in.

In my opinion, building aur packages in dependency order is the only safe way to upgrade.

**Comment by f2404** [comment]
@hauptmech Thanks for the suggestion! We will look into it.

**Comment by AladW** [comment]
Compare https://github.com/rmarquis/pacaur/issues/408

**Comment by z3ntu** [comment]
I think this comment belongs here:

We at the project "razer-drivers" had several reports / complaints over the last few months that installing our project with yaourt failed because of the following error:
```
loading packages...
resolving dependencies...
warning: cannot resolve "python-notify2", a dependency of "razer-daemon"
warning: cannot resolve "razer-daemon", a dependency of "openrazer-meta"
warning: cannot resolve "python-notify2", a dependency of "razer-daemon"
warning: cannot resolve "razer-daemon", a dependency of "python-razer"
warning: cannot resolve "python-razer", a dependency of "openrazer-meta"
warning: cannot resolve "python-notify2", a dependency of "razer-daemon"
warning: cannot resolve "razer-daemon", a dependency of "python-razer"
warning: cannot resolve "python-notify2", a dependency of "razer-daemon"
:: The following packages cannot be upgraded due to unresolvable dependencies:
      openrazer-meta  python-razer  razer-daemon
```

**Comment by AladW** [comment]
This is due to yet another aspect where yaourt is broken, being its inability to handle split packages. (not dicussed in this report) If you value your sanity, tell your users to install your project with makepkg or anything else with support for split packages.

**Comment by z3ntu** [comment]
Yes, we already have a warning on [our website](https://terrycain.github.io/razer-drivers/#arch) but we get still bug reports that the installation fails.

**Comment by AladW** [comment]
As much as I like to derail yaourt discussions I think reports you get for _your_ project are pretty far outside the scope of this project. Just add a rule to your spam filter that matches "yaourt" or "pacaur" to cover most bases.

**Comment by SolarAquarion** [comment]
@z3ntu @f2404 https://github.com/Jguer/yay/issues/10#issuecomment-352224998

see "aurchain"
